### PR TITLE
Run chef-client in a thread instead of a process (bnc#995027)

### PR DESCRIPTION
**Pull request by toabctl**
When the forked process (used to run the chef-client) exit, the database
connection is closed which leads to errors like:

PG::ConnectionBad: PQconsumeInput() server closed the connection unexpectedly

We currently catch this exception but we ran recently into something similar:

PG::UnableToSend: server closed the connection unexpectedly

https://bugzilla.suse.com/show_bug.cgi?id=995027

Hopefully this will be fixed when we use threads instead of processes.

### CODE REVIEW

**Comment by aplanas** [comment]
And use a connection pool? A helper that store N connections, and if the pool is empty, returns a new one.

**Comment by MaximilianMeister** [comment]
@toabctl good catch! thanks a lot for working on this, this might well cause the bug

maybe have a look at https://github.com/crowbar/crowbar-core/blob/291213be69ed3023274522c80d80334713416554/crowbar_framework/lib/crowbar/backup/restore.rb#L32 where we use the builtin rails threading, and see if that would work too.

File: crowbar_framework/app/models/service_object.rb [hunk 1]
```diff
@@ -1408,39 +1419,11 @@ def add_role_to_instance_and_node(barclamp, instance, name, prop, role, newrole)
...
-        rescue ::Exception
-          begin; STDOUT.reopen "/dev/null"; rescue ::Exception; end
-        end
-      else
-        begin; STDOUT.reopen "/dev/null"; rescue ::Exception; end
-      end
-      begin; STDERR.reopen STDOUT; rescue ::Exception; end
-      STDERR.sync = true
-
+    Thread.new {
```
Comment (MaximilianMeister) [inline crowbar_framework/app/models/service_object.rb]: ah i see you actually already use it here... my bad

**Comment by aspiers** [comment]
Sorry, I find the commit message a bit cryptic but maybe I'm just being stupid.  Which database connection is closed, in which process, and why?  And why would switching to threads help?

**Comment by nicolasbock** [comment]
Can you think of a way to test this hypothesis before we merge this? I don't understand why using threads as opposed to forking is helping.

**Comment by aplanas** [comment]
@nicolasbock @aspiers I can understand that if the process owner of the connection dies, the fd for the socket is closed. So if we use a single process with multiple threads, this fd is not closed when a thread ends.

But what I do not understand is why the owner of the connection is not the main process, but one of the child. This IMHO is a design problem. Hence the pool suggestion.

Also I concur with @nicolasbock to test it before : )

**Comment by nicolasbock** [comment]
Thanks @aplanas, now it's a little clearer. After thinking about this a bit I have this mental model now:
The main process is the owner of the file descriptor. It creates the DB connection before the `fork()`s and the children all get a duplicated `fd`. That `fd` is closed by a dying child which then of course affects the other children and the main process as well. Is that what you were thinking of?

**Comment by MaximilianMeister** [comment]
@aplanas @nicolasbock that's what we have gating for, just look at the logs for https://github.com/crowbar/crowbar-core/pull/582/files#diff-aa14186bf08406349d780bba0d7cfb72R28
or any other PG:: Error

I just found some links by googling "rails postgres process fork" that may help understanding the problem:
https://devcenter.heroku.com/articles/forked-pg-connections
http://stackoverflow.com/questions/27892975/what-causes-activerecord-breaking-postgres-connection-after-forking

**Comment by aplanas** [comment]
@nicolasbock I think that you nailed. Is the dup of the fd, that once is closed, break the connection for the rest. This explain the ownership problem.

@MaximilianMeister did you try the solution from heroku that you pointed? : disconnect before the fork and reconnect after that. Looks like that doing that the children will discover that there is no database connection, and will create a new one for this process.

**Comment by MaximilianMeister** [comment]
> did you try the solution from heroku that you pointed? : disconnect before the fork and reconnect after that. Looks like that doing that the children will discover that there is no database connection, and will create a new one for this process.

@aplanas no, I haven't tried it yet. but it seems that it doesn't create a new connection otherwise we wouldn't run into the errors... maybe it's a bug in https://github.com/rails/rails/blob/v4.2.0/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L623-625

and regarding the pools, we are already using 5 pools. see https://github.com/crowbar/crowbar-init/blob/master/chef/cookbooks/postgresql/templates/default/database.yml.erb#L22

btw. gating is green, and there are no errors or reconnects. maybe we can run a HA gating or sth similar with more parallel chef client runs?

**Comment by toabctl** [comment]
> And use a connection pool? A helper that store N connections, and if the pool is empty, returns a new one.

@aplanas I think Rails already uses a connection pool

**Comment by toabctl** [comment]
> Can you think of a way to test this hypothesis before we merge this?

@nicolasbock Yes. There's a easy reproducer for the first error message (found together with @rhafer yesterday). Do:

1) Remove the patch https://build.suse.de/package/view_file/Devel:Cloud:7:Staging/crowbar-core/582.patch?expand=1 from the package (or remove the exception handler in your env manually)
2) Call ```crowbar cinder proposal commit default
3) watch ```/var/log/crowbar/production.log``` .

This was 100% reproducable with the environment from https://ci.suse.de/view/Cloud/view/Cloud7/job/cloud-mkcloud7-job-magnum-x86_64/

For the second error message (```PG::UnableToSend: server closed the connection unexpectedly```) we haven't found a reproducer yet.

**Comment by toabctl** [comment]
Bw. this is another approach for #582 (which we have in our current package in D:C:7)

**Comment by toabctl** [comment]
There is some good documentation here: https://devcenter.heroku.com/articles/concurrency-and-database-connections#threaded-servers .

**Comment by toabctl** [comment]
> maybe have a look at https://github.com/crowbar/crowbar-core/blob/291213be69ed3023274522c80d80334713416554/crowbar_framework/lib/crowbar/backup/restore.rb#L32 where we use the builtin rails threading, and see if that would work too.

@MaximilianMeister What's the difference the code I use here? Both are using ```Thread.new```. No?

**Comment by MaximilianMeister** [comment]
> What's the difference the code I use here? Both are using Thread.new. No?

nevermind. I just thought on the first glance that you include another library for threading, but it seems that is just for the `wait for all threads`, which the `Thread` implementation doesn't provide
see https://github.com/crowbar/crowbar-core/pull/616#discussion_r75926943

**Comment by toabctl** [comment]
@MaximilianMeister yes. I'm using ```ThreadsWait``` which in the ruby standard libary.

File: crowbar_framework/app/models/service_object.rb [hunk 2]
```diff
@@ -1273,24 +1274,34 @@ def apply_role(role, inst, in_queue, bootstrap = false)
...
+        thread = run_remote_chef_client(node, "chef-client", filename)
+        threads[thread] = node
       end
-      status = Process.waitall
-      badones = status.select { |x| x[1].exitstatus != 0 }
 
-      next if badones.empty?
+      # wait for all running threads and collect the ones with a non-zero return value
+      bad_nodes = []
+      logger.debug "Waiting for #{threads.keys.length} threads to finish ..."
```
Comment (vuntz) [inline crowbar_framework/app/models/service_object.rb]: Detail, but no space before "..." :-)

File: crowbar_framework/app/models/service_object.rb (hunk 1 above)
Comment (rsalevsky) [inline crowbar_framework/app/models/service_object.rb]: @toabctl  Can you fix the hound issue? Just replace `Thread.new { ... }` with `Thread.new do ... end`.

File: crowbar_framework/app/models/service_object.rb (hunk 2 above)
Comment (toabctl) [inline crowbar_framework/app/models/service_object.rb]: Fixed

**Comment by nicolasbock** [comment]
+1

**Comment by rhafer** [comment]
> @MaximilianMeister did you try the solution from heroku that you pointed? : disconnect before the fork and reconnect after that. Looks like that doing that the children will discover that there is no database connection, and will create a new one for this process.

@aplanas We briefly tried that but as puma is multithreaded there are some issues with that approach. Closing all connections before the fork might close connections that currently have database transactions in-flight, which will cause other issues of it's own. Additionally closing and reopening all database connections of a pool (puma's pool is per process, i.e. all threads of a process share the same pool) for every fork (i.e. every chef-client run we trigger, plus some other forks) is pretty ugly as well.

**Comment by rhafer** [comment]
> For the second error message (PG::UnableToSend: server closed the connection unexpectedly) we haven't found a reproducer yet.

Actually I was able to see that error a few times now as well. It seems to be a timing issue of whether you get PG::UnableToSend or the other error. The underlying cause is the same (the child terminated the connection of the parent)

**Comment by aplanas** [comment]
@rhafer thanks for the details!!

Is amusing that the threaded approach worked in the gating. Moving from a non-shared variable space to a shared one is usually mode complex : )))))

If works: +1

**Comment by jsuchome** [comment]
I've met some problems with this approach. Did not look closely, but I've seen that the client chef log files that are supposed to be under /var/log/crowbar/chef-client were missing...

**Comment by MaximilianMeister** [comment]
> I've met some problems with this approach. Did not look closely, but I've seen that the client chef log files that are supposed to be under /var/log/crowbar/chef-client were missing...

I remember we had some issue with disappearing logs because of the forking. not sure if that applies here too. I just found 1 old commit now b534135ecee8896d90e6a93e1105c5ab3c581672 but I think @vuntz could remember more details

**Comment by rhafer** [comment]
> I remember we had some issue with disappearing logs because of the forking. not sure if that applies here too.

No, this is actually a problem of the patch. We're working on that.

**Comment by toabctl** [comment]
> I've met some problems with this approach. Did not look closely, but I've seen that the client chef log files that are supposed to be under /var/log/crowbar/chef-client were missing...

@jsuchome Fixed with the latest version...

**Comment by rhafer** [comment]
@aspiers  @toabctl My suggestion for the commit message (in the hope that it clears up some things):

When the forked process (used to execute the chef-client) terminates, it closes all its file descriptors. As it inherits all file descriptors of the parent process, that also causes the ActiveRecord's Database Connections to postgresql to be terminated. This also causes the parent process to loose the connection to the database.
The problems is visible in the log through errors like this:
"PG::ConnectionBad: PQconsumeInput() server closed the connection unexpectedly" or "PG::UnableToSend: server closed the connection unexpectedly" after the chef-client run finishes, when applying a proposal.

This patch tries to address the issue by avoiding the ```fork()``` call and using separate threads to launch chef-client. As a thread doesn't close any file descriptors when terminating. (Note we're still using the ```system``` call to launch chef-client, which obviously also forks another process, but as that is using some call of the ```exec``` family to run the command it does not mess around with the parent's file descriptors/sockets.

**Comment by toabctl** [comment]
@rhafer I updated the commit message. Thanks a lot!

**Comment by rhafer** [comment]
While I think the patch is good to address the immediate issue, it seems obvious to me that we need something more sophisticated in crowbar for running tasks in the background. @vuntz mentioned sucker_punch (https://github.com/brandonhilkert/sucker_punch) as one possible choice for that. (Apart from the name ;) )

**Comment by vuntz** [comment]
+1

**Comment by rsalevsky** [comment]
+1

**Pull request merged**